File: src/main/java/szewek/flux/FItems.java
package szewek.flux;

import net.minecraft.block.Block;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemStack;
import net.minecraftforge.registries.IForgeRegistry;
import szewek.flux.item.MetalItem;
import szewek.flux.item.MFToolItem;
import szewek.flux.item.GiftItem;
import szewek.flux.util.Metal;

import java.util.*;
import java.util.function.Function;
import java.util.function.Predicate;

import static szewek.flux.FluxMod.MODID;

public final class FItems {
	public static final ItemGroup MF_ITEMS = new ItemGroup("flux.items") {
		@Override
		public ItemStack createIcon() {
			return new ItemStack(FLUXGEN);
		}
	};

	public static final Map<Metal, MetalItem>
			GRITS = metalMap("grit", Metal::all),
			DUSTS = metalMap("dust", Metal::all),
			INGOTS = metalMap("ingot", Metal::nonVanilla);

	public static final MFToolItem MFTOOL = create(MFToolItem::new, "mftool", new Item.Properties().maxStackSize(1));
	public static final GiftItem GIFT = create(GiftItem::new, "gift", new Item.Properties().maxStackSize(1));
	public static final Item MACHINE_BASE = create(Item::new, "machine_base", new Item.Properties());
	public static final BlockItem
			FLUXGEN = fromBlock(FBlocks.FLUXGEN, "fluxgen"),
			GRINDING_MILL = fromBlock(FBlocks.GRINDING_MILL, "grinding_mill"),
			ALLOY_CASTER = fromBlock(FBlocks.ALLOY_CASTER, "alloy_caster"),
			WASHER = fromBlock(FBlocks.WASHER, "washer"),
			COMPACTOR = fromBlock(FBlocks.COMPACTOR, "compactor"),
			ENERGY_CABLE = fromBlock(FBlocks.ENERGY_CABLE, "energy_cable");

	static void register(final IForgeRegistry<Item> reg) {
		GRITS.values().forEach(reg::register);
		DUSTS.values().forEach(reg::register);
		INGOTS.values().forEach(reg::register);
		FBlocks.ORES.forEach((name, b) -> reg.register(fromBlock(b, name.name+"_ore")));
		FBlocks.METAL_BLOCKS.forEach((name, b) -> reg.register(fromBlock(b, name.name+"_block")));
		reg.registerAll(
				MFTOOL, GIFT, MACHINE_BASE,
				FLUXGEN, GRINDING_MILL, ALLOY_CASTER, WASHER, COMPACTOR, ENERGY_CABLE
		);
	}

	private static <T extends Item> T create(Function<Item.Properties, T> factory, String name, Item.Properties props) {
		T item = factory.apply(props.group(MF_ITEMS));
		item.setRegistryName(MODID, name);
		return item;
	}

	private static Map<Metal, MetalItem> metalMap(String type, Predicate<Metal> filter) {
		Map<Metal, MetalItem> m = new EnumMap<>(Metal.class);
		for (Metal metal : Metal.values()) {
			if (filter.test(metal)) {
				m.put(metal, create(MetalItem::new, metal.name + '_' + type, new Item.Properties()).withMetal(metal));
			}
		}
		return m;
	}

	private static BlockItem fromBlock(Block b, String name) {
		BlockItem item = new BlockItem(b, new Item.Properties().group(MF_ITEMS));
		item.setRegistryName(MODID, name);
		return item;
	}
}

File: src/main/java/szewek/flux/FItems.kt
package szewek.flux

import net.minecraft.block.Block
import net.minecraft.item.BlockItem
import net.minecraft.item.Item
import net.minecraft.item.ItemGroup
import net.minecraftforge.registries.IForgeRegistry
import szewek.flux.block.FluxOreBlock
import szewek.flux.block.MetalBlock
import szewek.flux.item.FluxToolItem
import szewek.flux.item.GiftItem
import szewek.flux.item.MetalItem
import szewek.flux.util.Metal
import szewek.ktutils.asStack
import java.util.*

object FItems {
    @JvmField val MF_ITEMS: ItemGroup = object : ItemGroup("flux.items") {
        override fun createIcon() = FLUXGEN.asStack
    }
    @JvmField val GRITS = metalMap("grit", Metal::always)
    @JvmField val DUSTS = metalMap("dust", Metal::always)
    @JvmField val INGOTS = metalMap("ingot", Metal::nonVanilla)
    @JvmField val FLUXTOOL = create(::FluxToolItem, "mftool", Item.Properties().maxStackSize(1))
    @JvmField val GIFT = create(::GiftItem, "gift", Item.Properties().maxStackSize(1))
    @JvmField val MACHINE_BASE = create(::Item, "machine_base", Item.Properties())
    @JvmField val FLUXGEN = fromBlock(FBlocks.FLUXGEN, "fluxgen")
    @JvmField val GRINDING_MILL = fromBlock(FBlocks.GRINDING_MILL, "grinding_mill")
    @JvmField val ALLOY_CASTER = fromBlock(FBlocks.ALLOY_CASTER, "alloy_caster")
    @JvmField val WASHER = fromBlock(FBlocks.WASHER, "washer")
    @JvmField val COMPACTOR = fromBlock(FBlocks.COMPACTOR, "compactor")
    @JvmField val ENERGY_CABLE = fromBlock(FBlocks.ENERGY_CABLE, "energy_cable")
    @JvmField val DIGGER = fromBlock(FBlocks.DIGGER, "digger")

    @JvmStatic
    fun register(reg: IForgeRegistry<Item?>) {
        GRITS.values.forEach(reg::register)
        DUSTS.values.forEach(reg::register)
        INGOTS.values.forEach(reg::register)
        FBlocks.ORES.forEach { (name: Metal, b: FluxOreBlock) -> reg.register(fromBlock(b, name.metalName + "_ore")) }
        FBlocks.METAL_BLOCKS.forEach { (name: Metal, b: MetalBlock) -> reg.register(fromBlock(b, name.metalName + "_block")) }
        reg.registerAll(
                FLUXTOOL, GIFT, MACHINE_BASE,
                FLUXGEN, GRINDING_MILL, ALLOY_CASTER, WASHER, COMPACTOR, ENERGY_CABLE, DIGGER
        )
    }

    private fun <T : Item?> create(factory: (Item.Properties) -> T, name: String, props: Item.Properties): T {
        val item = factory.invoke(props.group(MF_ITEMS))
        item!!.setRegistryName(FluxMod.MODID, name)
        return item
    }

    private fun metalMap(type: String, filter: (Metal) -> Boolean): Map<Metal, MetalItem> {
        val m: MutableMap<Metal, MetalItem> = EnumMap(Metal::class.java)
        for (metal in Metal.values()) {
            if (filter.invoke(metal)) {
                m[metal] = create(::MetalItem, metal.metalName + '_' + type, Item.Properties()).withMetal(metal)
            }
        }
        return m
    }

    private fun fromBlock(b: Block, name: String): BlockItem {
        val item = BlockItem(b, Item.Properties().group(MF_ITEMS))
        item.setRegistryName(FluxMod.MODID, name)
        return item
    }
}